src/application/service/CreateOrUpdateTest.ts

```diff
@@ -1,33 +1,58 @@
 import { UserId } from '../../domain/model/user/UserId';
 import { Test } from '../../domain/model/test/Test';
 import { TestId } from '../../domain/model/test/TestId';
+import { Results } from '../../domain/model/test/Results';
 import { TestRepository } from '../../domain/model/test/TestRepository';
 import { getTestTypes } from './index';
 import { TestTypeId } from '../../domain/model/testType/TestTypeId';
-import { Test as ApiTest } from '../../api/interface/index';
+import { TestCommand } from '../../api/interface/index';
+import Ajv from 'ajv';
+
+const jsonSchema = new Ajv({
+  strictDefaults: true,
+  strictKeywords: true,
+});
+
+import { DomainValidationError } from '../../domain/model/DomainValidationError';
+import { ResourceNotFoundError } from '../../domain/model/ResourceNotFoundError';
 
 export class CreateOrUpdateTest {
   constructor(private testRepository: TestRepository) {}
 
-  public async execute(userId: string, payload: ApiTest): Promise<Test> {
+  public async execute(userId: string, payload: TestCommand): Promise<Test> {
     const { testTypeId, results } = payload;
-    const checkTestType = await getTestTypes.byId(testTypeId);
+    const testType = await getTestTypes.byId(testTypeId);
+
+    if (!testType) {
+      throw new ResourceNotFoundError('testTypeId', testTypeId);
+    }
+
+    if (results) {
+      const isValid = jsonSchema.validate(
```

Comment:
Sorry I missed this before  but wouldn't it be better to move this validation to the domain? 

I also talked briefly about this with Harsh at the time. 

We can introduce a TestResultsFactory which can be instantiated with a testTypeRepository. 

Then TestResultFactory can have a TestFactory#createTestResults which only ever creates valid Results objects. 

This way this code can be moved to the application and domain package and we will guarantee that there is no way to create an invalid TestResult

```

```